src/main/java/edu/kpi/testcourse/storage/UserRepositoryFileImpl.java

```diff
@@ -0,0 +1,86 @@
+package edu.kpi.testcourse.storage;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import edu.kpi.testcourse.entities.User;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * An file-backed implementation of {@link UserRepository} suitable for use in production.
+ */
+public class UserRepositoryFileImpl implements UserRepository {
+  private final Map<String, User> users;
+  private final Gson gson;
+  private final String storageRoot;
+
+  /**
+   * Creates an instance for production.
+   */
+  @Inject
+  public UserRepositoryFileImpl(Gson gson) {
+    this.gson = gson;
+    this.storageRoot = "/tmp/url-shortener-db";
+    this.users = readFromFile();
+  }
+
+  /**
+   * Creates an instance for use in unit tests.
+   */
+  UserRepositoryFileImpl(Gson gson, String storageRoot) {
+    this.gson = gson;
+    this.storageRoot = storageRoot;
+    this.users = readFromFile();
+  }
+
+  @Override
+  public void createUser(User user) {
+    if (users.putIfAbsent(user.email(), user) != null) {
+      throw new RuntimeException("User already exists");
+    }
+    writeToFile();
+  }
+
+  @Override
+  public @Nullable User findUser(String email) {
+    return users.get(email);
+  }
+
+  private Path getJsonFilePath() {
+    return Paths.get(storageRoot + "/user-repository.json");
```

Comment:
In this code, we use raw strings as paths, together with the `Path` class. Why don't eliminate raw strings entirely in the codebase, and use `Path` class that will save us from concatenation and os specific errors?